packages/webapp/src/components/Form/CompositionInputs/NumberInputWithSelect.tsx

```diff
@@ -0,0 +1,137 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+import { ReactNode, useEffect } from 'react';
+import clsx from 'clsx';
+import { useTranslation } from 'react-i18next';
+import ReactSelect from '../ReactSelect';
+import useReactSelectStyles from '../Unit/useReactSelectStyles';
+import useNumberInput from '../NumberInput/useNumberInput';
+import InputBase from '../InputBase';
+import { styles as reactSelectDefaultStyles } from '../ReactSelect';
+import styles from './styles.module.scss';
+
+type Option = { value: string; label: ReactNode };
+
+export const UNIT = 'unit';
+
+export enum NPK {
+  N = 'n',
+  P = 'p',
+  K = 'k',
+}
+
+export type NumberInputWithSelectProps = {
+  name: NPK;
+  label: string;
+  unitOptions: Option[];
+  disabled?: boolean;
+  error?: string;
+  className?: string;
+  values?: {
+    [NPK.N]?: number | null;
+    [NPK.P]?: number | null;
+    [NPK.K]?: number | null;
+    [UNIT]?: Option;
+  };
+  onChange: (fieldName: string, value: number | Option | null) => void;
+};
+
+const REACT_SELECT_WIDTH = 44;
+
+const NumberInputWithSelect = ({
+  name,
+  label,
+  unitOptions,
+  disabled,
+  error,
+  className,
+  onChange,
+  values = {},
+}: NumberInputWithSelectProps) => {
+  const { t } = useTranslation();
+  const unit = values[UNIT];
+
+  const reactSelectStyles = useReactSelectStyles(disabled, {
+    reactSelectWidth: REACT_SELECT_WIDTH,
+  });
+  reactSelectStyles.control = (provided) => ({
+    ...provided,
+    boxShadow: 'none',
+    borderRadius: 0,
+    height: '46px',
+    paddingLeft: disabled ? '4px' : '8px',
+    fontSize: '16px',
+    border: 'none',
+    background: 'inherit',
+  });
+  reactSelectStyles.singleValue = (provided) => ({
+    ...provided,
+    color: 'var(--Colors-Neutral-Neutral-300, #98A1B1);',
+    paddingTop: unit?.value === 'ratio' ? '4px' : 0,
+  });
+  reactSelectStyles.option = (provided, state) => ({
+    ...reactSelectDefaultStyles.option?.(provided, state),
+    color: 'var(--Colors-Neutral-Neutral-300, #98A1B1);',
+  });
+
+  const { inputProps, update } = useNumberInput({
+    onChange: (value) => onChange(name, value),
+    initialValue: values[name],
+    max: 999999999,
+  });
+
+  useEffect(() => {
+    // If the value is updated from the parent, update the actual value in the input
+    if (inputProps.value !== values[name]) {
```

Comment:
I think this is causing an issue in the `With Error` story. I'm not able to type the value its updating it as soon as I type.

Another potential issue is that inputProps.value could be a string and value[name] is a number so that might cause issues when comparing.

Reply:
Thank you for catching!!!

Do you think it will work if we don't compare their types?

Reply:
Sorry I should've mentioned this before but this won't work if the number is localized. For example on the 'pt' locale the number 23.5 would be 23,5 and 23,5 == 23.5 would be false. I'm just wondering is there a need for the useEffect here? If its really needed then `numericValue` can probably be used instead of inputProps.value

Reply:
Oh that's right...
So the problematic flow is editing an existing product.
If the existing product has N: 20 and the user changes it to 30 and then click "cancel", the value needs to be back to "20". I will use `reset` returned from `useForm`, and it makes the value "20" and it needs to be reflected in the `useNumberInput`.
`numericValue` sounds right to use, I'll try, thank you!

Reply:
Oh ok that makes sense. 

I remember reading about how adjusting the state with an effect might be a bad idea in the react docs - here's the link https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes

They mention how its more performant to do it without the effect, just wondering if that could be applied here?

Reply:
Thank you so much for sharing the link, I didn't know useEffect **should be your last resort** 😰

I tried removing `useEffect` and tested using a component I'm working on in my next [ticket](https://lite-farm.atlassian.net/browse/LF-4220) and got this error:
 
![Screenshot 2024-05-09 at 8 31 35 AM](https://github.com/LiteFarmOrg/LiteFarm/assets/33141219/3dd62056-1d78-424d-b136-6c7463c86f9e)

It seems to be the case where `useEffect` is a solution according to this [React repo issue](https://github.com/facebook/react/issues/18178).

It would be appreciated if you could check the fixed condition!